### PR TITLE
HAL_Linux: ignore return value of writing lf

### PR DESCRIPTION
**Pull request by peterbarker**

### CODE REVIEW

File: libraries/AP_HAL_Linux/system.cpp [hunk 1]
```diff
@@ -27,7 +27,9 @@ void panic(const char *errormsg, ...)
     va_start(ap, errormsg);
     vdprintf(1, errormsg, ap);
     va_end(ap);
-    write(1, "\n", 1);
+    if (write(1, "\n", 1) != 1) {
```
Comment (guludo) [inline libraries/AP_HAL_Linux/system.cpp]: I think a more correct approach is using pragmas:
```c++
    #pragma GCC diagnostic push
    #pragma GCC diagnostic ignored "-Wunused-result"
    write(1, "\n", 1);
    #pragma GCC diagnostic pop
```

... and it works for both GCC and clang. I know that is more verbose than using `if`, but I think it's makes more sense.

**Comment by peterbarker** [comment]
@guludo This one came up at the dev call.  I prefer the non-#pragma approach here to keep the visual noise down, and people seemed to prefer neither one nor the other.

I'm curious - why your preference for the #pragma-based approach?

**Comment by magicrub** [comment]
for the record, I'm in favor of the current implementation due to low-noise
impact.

**Comment by guludo** [comment]
> I'm curious - why your preference for the #pragma-based approach?

@peterbarker Just because using an `if` statement seems an awkward way to prevent that warning. I mean, `if` statement isn't intended for that. Using pragmas seems more correct semantically speaking, IMHO. However I agree that it is too much verbose.

I was just doing some tests here and I think doing the following could work:
```c++
int r = write(1, "\n", 1);
(void)r;
```
But I still don't like it hehehe


My idea was actually to do the following:
```c++
#define IGNORE_RESULT(expr_) \
    _Pragma("GCC diagnostic push") \
    _Pragma("GCC diagnostic ignored \"-Wunused-result\"") \
    expr_; \
    _Pragma("GCC diagnostic pop")
```

So that we could use `IGNORE_RESULT(write(1, "\n", 1))`. That works with clang, but not in gcc due to a bug (here: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=55578).

Well, I'm in favor of defining a macro `IGNORE_RESULT`, which could be changed in the future with no problems. It could be defined as:

```c++
#define IGNORE_RESULT(expr_) do { if (expr_) {} } while(0)
```

The `while` block is only for forcing the use of `;` at the end of `IGNORE_RESULT()`. Using that macro makes the code more reasonable than just an explicit `if` statement where the result should be ignored. That also allows us to change how to ignore the result of a function with the attribute `warn_unused_result` in just one place (e.g., when that bug in GCC is fixed).

**Comment by lucasdemarchi** [comment]
(void) is the standard way to ignore the result, no need for pragmas.... btw you can just put (void) in front of the function call instead of creating a variable.

the if approach would be good if we actually did anything in the branch rather than leaving it empty.

**Comment by OXINARF** [comment]
@lucasdemarchi I think @tridge said in the dev call that was removed in GCC 6.

**Comment by lucasdemarchi** [comment]
I tested recently and it was working.

**Comment by guludo** [comment]
```c++
int foo() __attribute__(( warn_unused_result ));

int foo()
{
    return 0;
}

int main()
{
    (void)foo();
    return 0;
}
```

GCC 6.1.1 will still raise a warning for this code.

**Comment by lucasdemarchi** [comment]
Just remembered that I commented in a GCC bug some time ago....
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=66425
Em 13 de jun de 2016 3:18 PM, "Gustavo José de Sousa" <
notifications@github.com> escreveu:

**Comment by dgrat** [comment]
I also prefer a macro to ignore the result. This is btw also the libraries like Qt handle this problem.

**Comment by peterbarker** [comment]
Outcome from the meeting today was to use the macro @guludo mentioned above.

That makes this PR obsolete.  Seems to have served a purpose, 'though!

**Pull request closed without merging**